### PR TITLE
misc: use VarConstraint instead of Annotated[ConstraintVar] in operation definitions

### PR DESCRIPTION
**Pull request by superlopuh**
Using Annotated in the way that we do is not well defined by the Python type system, this fixes the current misuse to get the number of Pylance errors to 0.

The current proposal is to keep both implementations working, and to deprecate the `Annotated` API at some point in the future.

### CODE REVIEW

File: tests/irdl/test_declarative_assembly_format.py [hunk 1]
```diff
@@ -1661,13 +1664,25 @@ class ParamOne(ParametrizedAttribute, TypeAttribute, Generic[_T]):
         p: ParameterDef[_T]
         q: ParameterDef[Attribute]
 
+        @classmethod
```
Comment (alexarice) [inline tests/irdl/test_declarative_assembly_format.py:1667]: This seems overkill for this test?

File: tests/irdl/test_operation_definition.py [hunk 2]
```diff
@@ -227,6 +230,78 @@ def test_constraint_var_fail_not_satisfy_constraint():
         op.verify()
 
 
+@irdl_op_definition
+class GenericConstraintVarOp(IRDLOperation):
+    name = "test.constraint_var_op"
+
+    T: ClassVar[VarConstraint[IntegerType | IndexType]] = VarConstraint(
+        "T", base(IntegerType) | base(IndexType)
+    )
```
Comment (alexarice) [inline tests/irdl/test_operation_definition.py:239]: In general do we want these constraints to be a visible part of the operation interface? Would it be better to use `_T` instead?

File: tests/irdl/test_operation_definition.py [hunk 3]
```diff
@@ -227,6 +230,78 @@ def test_constraint_var_fail_not_satisfy_constraint():
...
+def test_generic_constraint_var_fail_non_equal():
+    """Check that all uses of a constraint variable are of the same attribute."""
+    i32_operand = TestSSAValue(i32)
+    index_operand = TestSSAValue(IndexType())
+
+    # Fail because of operand
+    op = GenericConstraintVarOp.create(
+        operands=[index_operand], result_types=[i32], attributes={"attribute": i32}
+    )
+    with pytest.raises(DiagnosticException):
```
Comment (alexarice) [inline tests/irdl/test_operation_definition.py:271]: It would be nice for these to match on the error

File: tests/tblgen_to_py/test.py [hunk 4]
```diff
@@ -50,13 +50,7 @@ class Test_AttributesOp(IRDLOperation):
...
 
     int_attr = prop_def(
-        ParamAttrConstraint(
-            IntegerAttr,
-            (
-                AnyAttr(),
-                EqAttrConstraint(IntegerType(16)),
-            ),
-        )
+        IntegerAttr[IntegerType].constr(type=EqAttrConstraint(IntegerType(16)))
```
Comment (alexarice) [inline tests/tblgen_to_py/test.py:53]: This file was automatically generated and probably shouldn't be manually changed

File: xdsl/dialects/builtin.py [hunk 5]
```diff
@@ -644,6 +665,7 @@ def __init__(
 
 
 AnyFloatAttr: TypeAlias = FloatAttr[AnyFloat]
+AnyFloatAttrConstr: BaseAttr[AnyFloatAttr] = BaseAttr(FloatAttr)
```
Comment (alexarice) [inline xdsl/dialects/builtin.py:668]: Do we need to put the type here?

File: xdsl/dialects/ltl.py [hunk 6]
```diff
@@ -51,11 +51,9 @@ class AndOp(IRDLOperation):
 
     name = "ltl.and"
 
-    T = Annotated[
-        Attribute,
-        AnyOf([Sequence, Property, IntegerType(1, signedness=Signedness.SIGNLESS)]),
-        ConstraintVar("T"),
-    ]
+    T: ClassVar[VarConstraint[Attribute]] = VarConstraint(
+        "T", AnyOf([Sequence, Property, IntegerType(1, signedness=Signedness.SIGNLESS)])
```
Comment (alexarice) [inline xdsl/dialects/ltl.py:55]: `Signedness.SINGLESS` is the default I believe

File: xdsl/irdl/constraints.py [hunk 7]
```diff
@@ -221,6 +221,10 @@ def verify(
         pass
 
 
+# TODO: merge after pyright is updated, and update all constrs
+ANY_ATTR = AnyAttr()
```
Comment (alexarice) [inline xdsl/irdl/constraints.py:225]: Why is this necessary?

File: xdsl/tools/tblgen_to_py.py [hunk 8]
```diff
@@ -319,16 +319,12 @@ def _resolve_prop_constraint(self, rec: TblgenRecord | str) -> str:
                 return "BaseAttr(BoolAttr)"
             case "IndexAttr":
                 return textwrap.dedent("""
-                ParamAttrConstraint(
-                    IntegerAttr, (AnyAttr(), EqAttrConstraint(IndexType()))
-                )
+                IntegerAttr[IndexType].constr(type=IndexTypeConstr)
```
Comment (alexarice) [inline xdsl/tools/tblgen_to_py.py:322]: Nice, ignore the earlier comment

**Review by alexarice — COMMENTED** [review]
Very nice. Not a comprehensive review but I left some comments. Is the `constr` field a part of the `ParametrizedAttr` superclass?

File: xdsl/irdl/constraints.py (hunk 7 above)
Comment (superlopuh) [reply xdsl/irdl/constraints.py:225]: It's not really, it's more of a perf thing, just feels wrong to me to create lots of the same object, but I'm not planning on including it in any of the upcoming PRs, at least not before updating Pyright.

File: tests/irdl/test_operation_definition.py (hunk 2 above)
Comment (superlopuh) [reply tests/irdl/test_operation_definition.py:239]: What `_T`?

File: tests/irdl/test_declarative_assembly_format.py [hunk 9]
```diff
@@ -1678,11 +1678,11 @@ def constr(
 
     @irdl_op_definition
     class TwoOperandsNestedVarOp(IRDLOperation):
-        T = Annotated[Attribute, ConstraintVar("T")]
+        T: ClassVar[VarConstraint[Attribute]] = VarConstraint("T", AnyAttr())
 
         name = "test.two_operands_one_result_with_var"
         res = result_def(T)
-        lhs = operand_def(ParamOne[T])
+        lhs = operand_def(ParamOne[Attribute].constr(p=T))
```
Comment (math-fehr) [inline tests/irdl/test_declarative_assembly_format.py:1685]: I do have to say, I really dislike the verbosity of this, it will make it harder for newcomers to get started I think.

**Review by math-fehr — APPROVED** [review]
🚢

File: tests/irdl/test_declarative_assembly_format.py (hunk 9 above)
Comment (superlopuh) [reply tests/irdl/test_declarative_assembly_format.py:1685]: Yeah it definitely got less pretty. One potential alternative is `ParamOneConstr(p=T)`, would you like me to try that instead? It's one more thing to import, but potentially worth the effort.

Comment (superlopuh) [reply tests/irdl/test_declarative_assembly_format.py:1685]: I just tested it locally and that API works and makes Pyright happy, it really doesn't want you to omit the type parameter even if it could be inferred in this case. I'm still inclined towards this more verbose API for the time being, with the hope that Pyright gets smart enough at some point to drop the unnecessary type parameter. We could also file it as a bug.

Comment (superlopuh) [reply tests/irdl/test_declarative_assembly_format.py:1685]: https://github.com/microsoft/pyright/issues/9190

Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:1685]: What about making `constr` a staticmethod instead? I just tried it locally and my pyright at least seems happy?

Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:1685]: This allows you to do `ParamOne.constr(p=T)` instead of `ParamOne[Attribute].constr(p=T)`

File: tests/irdl/test_operation_definition.py (hunk 2 above)
Comment (superlopuh) [reply tests/irdl/test_operation_definition.py:251]: Oh yeah I guess they could be, I tried to modify a minimal amount with these changes.

File: tests/irdl/test_declarative_assembly_format.py (hunk 9 above)
Comment (superlopuh) [reply tests/irdl/test_declarative_assembly_format.py:1685]: it doesn't work for me

Comment (superlopuh) [reply tests/irdl/test_declarative_assembly_format.py:1685]: How do you feel about iterating on this after this is merged and Pyright is updated?

Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:1685]: Fine with me

Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:1685]: https://peps.python.org/pep-0696/ also potentially relevant for making this api nicer

Comment (superlopuh) [reply tests/irdl/test_declarative_assembly_format.py:1685]: Yep, I look forward to being able to use the new, shiny APIs. I'm not sure when it would be wise to drop support for older Pythons, according to this calendar 3.10 is supported for another couple of years: https://endoflife.date/python
